js/tests/unit/util/index.spec.js

```diff
@@ -521,10 +521,10 @@ describe('Util', () => {
 
     it('should execute if arg is function & return the result', () => {
       const functionFoo = (num1, num2 = 10) => num1 + num2
-      const resultFoo = Util.execute(functionFoo, [4, 5])
+      const resultFoo = Util.execute(functionFoo, [undefined, 4, 5])
```

Comment:
I'm in the process of understanding what was there before and what's changing in this PR. It's a bit challenging to think about all the use cases, so I hope you won't mind if my remarks or questions are not yet precise or accurate.

If we don't think about the entire context of this PR and focus on this `Util.execute` function only, I'm not that comfortable with an extra first parameter that seems useless. Considering this function in isolation, we would probably expect it to work like this:
* `Util.execute(<function reference>)` or `Util.execute(<function reference>, [])` when there's no args
* `Util.execute(<function reference>, [<arg 1>]`
* `Util.execute(<function reference>, [<arg 1>, <arg 2>]`
* etc.

I'm not saying it doesn't have impacts, or is feasible technically, but in terms of pure interface don't you think this `undefined` first argument in the list should be removed?

Reply:
I think my rationale for doing it this way is that the `args` array directly mirrors how `Function.prototype.call` works, where the first argument is `this` and the remaining ones are the arguments.

I'm entirely happy to change the interface, e.g. by updating `Util.execute` to explicitly take a value to be used at `this`. For instance, we could update `Util.execute` to look like this:

```ts
const execute = (possibleCallback, thisArg = undefined, args = [], defaultValue = possibleCallback)
```

Then, the test case you commented on here would look like this:

```ts
const resultFoo = Util.execute(functionFoo, undefined, [4, 5]);
```

Of course, we could reorder the arguments, but I think putting it before the arguments (again to mirror `Function.prototype.call`) would make the most sense.

It's also possible that `Util.execute` is simply the wrong thing to be using in the callbacks for tooltips. As mentioned in https://github.com/twbs/bootstrap/pull/38725#discussion_r1255005202, I'm happy to switch those callsites to essentially inline it instead:

```js
typeof arg === 'function' ? arg.call(this._element) : arg

// If we want to retain backwards compatibility with the new API accidentally
// introduced in 5.3, we can do this instead:
typeof arg === 'function' ? arg.call(this._element, this._element) : arg
```

Reply:
It wasn't immediately clear to me, but now I see the value in mirroring `Function.prototype.call` as you suggested.

> `const resultFoo = Util.execute(functionFoo, undefined, [4, 5]);`

With your explanation, I'd say that your version in this PR, `Util.execute(functionFoo, [undefined, 4, 5]);` is a better fit because it aligns more closely with `Function.prototype.call`.

> [...] but I think putting it before the arguments (again to mirror Function.prototype.call) would make the most sense.

Yes it would.

> It's also possible that Util.execute is simply the wrong thing to be using in the callbacks for tooltips. As mentioned in https://github.com/twbs/bootstrap/pull/38725#discussion_r1255005202, I'm happy to switch those callsites to essentially inline it instead

I'm not the sole decision-maker here, but from my perspective as a non-JS expert, either approach could work. The initial idea with https://github.com/twbs/bootstrap/pull/36652 was to consolidate functionality and avoid repetitive `typeof arg === 'function'` checks. Thus, retaining `execute` seems reasonable if we want to maintain that approach.

You've already done the work to keep `execute`, so given that I don't have a strong preference at the moment, I'll continue to review the use cases from this angle.

_Note: I'm currently trying to reach the JS maintainers to discuss how we can proceed and finalize this PR with their input and yours._

Reply:
Sounds great! I'll be ready to chat with you or the other JS maintainers to find an approach that'll make everyone happy.

Just so it isn't overlooked, a main decision to make here is whether or not Bootstrap will continue calling these functions with the tooltip/popover/etc. as an explicit argument _as well as_ the old behavior of setting `this`. If there's a desire to keep the accidentally-introduced new behavior, I'll add tests and update docs to reflect that.

Reply:
OK, so we got a regression where folks had to use a workaround. The basic principle of a fix is that when there's a new patch release (v5.3.x), folks can then remove their workaround so that it works like before.

Since we didn't release a patch promptly, this workaround might now be seen as a de facto part of the API. Even if it isn't officially recognized as such, we may decide not to inconvenience users who have relied on and adapted to this workaround for an extended period. Otherwise, they might perceive the new patch as a disruptive change/breaking change as well, if not retro-compatible. 🙃

Considering these factors, I'd say supporting both the old and the new behavior would ensure a seamless transition for everyone installing the patch version (and we might even consider later on deprecating the old behavior in a future major Bootstrap version, for example):
- Someone migrating from v5.1 to this new v5.3.x could use everything as described in the documentation
- Someone who had to use a workaround could install the new v5.3.x transparently

Yet, I'm wondering whether introducing the new behavior in the documentation should wait until v5.4.0. Even though the feature could be currently present but hidden, delaying its documentation could help us avoid unnecessary remarks and comments. Not entirely sure if this is the right strategy.

Reply:
Maybe a reasonable middle ground is to implement+test both but avoid documenting the "new" behavior until either a future minor/major release?

Reply:
Yeah, that was the idea I had in mind.

1. Implement + test both so that it works for all use cases (migrating from versions before the breaking change, but also supporting transparently workarounds that have been applied). It would give us some space if there are mistakes with the new implementation (as long as the old implementation is fixed and the previous doc still works).

2. Once we have ensured stability, we can document and present the "new" behavior as a minor feature in v5.4.

If this approach sounds reasonable and feasible to you and others involved in this PR, I suggest we proceed with this strategy in mind.

Reply:
I updated the implementation and tests in a1a57a0.